packages/lib/gpcircuits/src/multituple.ts

```diff
@@ -0,0 +1,232 @@
+    throw new RangeError("The tuple arity parameter must be at least 2.");
+  } else if (paramMaxTuples < 0) {
+    throw new RangeError(
+      "The maximum tuple number parameter must be nonnegative."
+    );
+  } else {
+    return paramMaxTuples === 0
+      ? // If we have no tuples at our disposal, we cannot represent any tuples,
+        // thus return arity 0.
+        0
+      : // Else we have `paramTupleArity` slots in the first tuple at our
+        // disposal and `paramTupleArity` - 1 slots in every subsequent tuple,
+        // since the first slot of these tuples will contain the index pointing
+        // to the preceding tuple.
+        paramTupleArity + (paramTupleArity - 1) * (paramMaxTuples - 1);
+  }
+}
+
+/**
+ * This procedure takes an N-tuple of `indices` (referring
+ * to elements of some array of values) and returns its
+ * representation as an appropriately linked sequence of
+ * tuples of arity `paramMaxArity`, where these tuples
+ * are indexed starting at `firstAvailableTupleIndex` and the
+ * first index is used as padding. 'Appropriately linked'
+ * means that the first slot of every output tuple after
+ * the first has an index referring to the preceding
+ * tuple.
+ *
+ * Examples:
+ * computeTupleIndices(2, 5, [1, 3, 4]) === [[1, 3], [5, 4]]
+ * computeTupleIndices(3, 5, [0, 1, 4, 2]) === [[0, 1, 4], [5, 2, 0]]
+ * computeTupleIndices(4, 6, [3, 4, 2, 1, 5]) === [[3, 4, 2, 1], [6, 5, 3, 3]]
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param firstAvailableTupleIndex the index of the first output tuple in the combined array of
+ * entry and tuple value hashes
+ * @param indices the entry value indices forming the input tuple
+ * @returns list of index tuples of arity `paramTupleArity` representing the input tuple
+ * @throws RangeError if the parameters are not in the proper ranges.
+ */
+export function computeTupleIndices(
+  paramTupleArity: number,
+  firstAvailableTupleIndex: number,
+  indices: number[]
+): number[][] {
+  if (indices.some((i) => i >= firstAvailableTupleIndex)) {
+    throw new RangeError(
+      "The first available tuple index must be strictly greater than the given indices."
+    );
+  }
+
+  // Note the first index.
+  const firstIndex = indices[0];
+
+  // The tuple indices will begin with the first `paramTupleArity` indices.
+  const tupleIndices = [indices.slice(0, paramTupleArity)];
+
+  // If we have more than that,
+  if (indices.length > paramTupleArity) {
+    // split the remainder into `paramTupleArity - 1` sized chunks.
+    const unlinkedChunks = toChunks(
+      indices.slice(paramTupleArity),
+      paramTupleArity - 1
+    );
+
+    // Prepend the 'index' of the preceding tuple
+    // to each of these chunks and push them onto `tupleIndices`.
+    unlinkedChunks.forEach((chunk, i) =>
+      tupleIndices.push([firstAvailableTupleIndex + i].concat(chunk))
+    );
+  }
+
+  // Take note of the last tuple for padding purposes.
```

Comment:
These variables, plus firstIndex above, could all fall below the single comment on the loop below.  That comment explains the purpose, and the variable names are very clear on each value's purpose without the need for extra comments about "noting" the values.